app/controllers/account/permissions_controller.rb

```diff
@@ -14,6 +14,15 @@ def show
 
   def edit
     authorize [:account, @application], :edit_permissions?
+
+    @permissions = @application.sorted_supported_permissions_grantable_from_ui
+  end
+
+  def update
+    authorize [:account, @application], :edit_permissions?
+
+    current_user.replace_application_permissions(@application, params[:application][:permissions])
```

Comment:
I think we need to use `UserUpdate` here because that already handles:

- adding/removing only the permissions that the user is allowed to modify (through the [`SupportedPermissionParameterFilter`](https://github.com/alphagov/signon/blob/de64c49d9017518421ad0844ef4ec811068a2c57/app/services/user_update.rb#L35))
- logging the addition/removal of permissions

I guess you'll need to merge all existing permissions with the permissions that the user has added/removed through the UI for this particular application. The [`Account::SigninPermissionsController#create`](https://github.com/alphagov/signon/blob/de64c49d9017518421ad0844ef4ec811068a2c57/app/controllers/account/signin_permissions_controller.rb#L9-L10) is doing something similar so might give you some clues.

Reply:
@chrisroos and I have paired on this and addressed this issue in [a524c11](https://github.com/alphagov/signon/pull/2488/commits/a524c11ff7ea94d980c29d74f6cdb0047e695a14)